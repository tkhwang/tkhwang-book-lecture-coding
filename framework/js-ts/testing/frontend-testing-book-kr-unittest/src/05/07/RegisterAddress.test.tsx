import { render, screen } from "@testing-library/react";
import { mockPostMyAddress } from "./fetchers/mock";
import { RegisterAddress } from "./RegisterAddress";
import {
  clickSubmit,
  inputContactNumber,
  inputDeliveryAddress,
} from "./testingUtils";

jest.mock("./fetchers");

async function fillValuesAndSubmit() {
  const contactNumber = await inputContactNumber();
  const deliveryAddress = await inputDeliveryAddress();
  const submitValues = { ...contactNumber, ...deliveryAddress };
  await clickSubmit();
  return submitValues;
}

async function fillInvalidValuesAndSubmit() {
  const contactNumber = await inputContactNumber({
    name: "배언수",
    phoneNumber: "abc-defg-hijkl",
  });
  const deliveryAddress = await inputDeliveryAddress();
  const submitValues = { ...contactNumber, ...deliveryAddress };
  await clickSubmit();
  return submitValues;
}

beforeEach(() => {
  jest.resetAllMocks();
});

/* 코드 5-49
test("유효성 검사 에러가 발생하면 메시지가 표시된다", async () => {
  render(<RegisterAddress />);
  await fillInvalidValuesAndSubmit();
  expect(screen.getByText("올바르지 않은 값이 포함되어 있습니다")).toBeInTheDocument();
});
*/

/* 코드 5-50
test("원인이 명확하지 않은 에러가 발생하면 메시지가 표시된다", async () => {
  render(<RegisterAddress />);
  await fillValuesAndSubmit();
  expect(screen.getByText("알 수 없는 에러가 발생했습니다")).toBeInTheDocument();
});
*/

test("성공하면 '등록됐습니다'가 표시된다", async () => {
  const mockFn = mockPostMyAddress();
  render(<RegisterAddress />);
  const submitValues = await fillValuesAndSubmit();
  expect(mockFn).toHaveBeenCalledWith(expect.objectContaining(submitValues));
  expect(screen.getByText("등록됐습니다")).toBeInTheDocument();
});

test("실패하면 '등록에 실패했습니다'가 표시된다", async () => {
  const mockFn = mockPostMyAddress(500);
  render(<RegisterAddress />);
  const submitValues = await fillValuesAndSubmit();
  expect(mockFn).toHaveBeenCalledWith(expect.objectContaining(submitValues));
  expect(screen.getByText("등록에 실패했습니다")).toBeInTheDocument();
});

test("유효성 검사 에러가 발생하면 '올바르지 않은 값이 포함되어 있습니다'가 표시된다", async () => {
  render(<RegisterAddress />);
  await fillInvalidValuesAndSubmit();
  expect(screen.getByText("올바르지 않은 값이 포함되어 있습니다")).toBeInTheDocument();
});

test("원인이 명확하지 않은 에러가 발생하면 '알 수 없는 에러가 발생했습니다'가 표시된다", async () => {
  render(<RegisterAddress />);
  await fillValuesAndSubmit();
  expect(screen.getByText("알 수 없는 에러가 발생했습니다")).toBeInTheDocument();
});

test("Snapshot: 등록 폼이 표시된다", async () => {
  mockPostMyAddress();
  // const mockFn = mockPostMyAddress();
  const { container } = render(<RegisterAddress />);
  // const submitValues = await fillValuesAndSubmit();
  // expect(mockFn).toHaveBeenCalledWith(expect.objectContaining(submitValues));
  expect(container).toMatchSnapshot();
});

/* 코드 5-55
test("Snapshot: 등록 폼이 표시된다", async () => {
  // mockPostMyAddress();
  const mockFn = mockPostMyAddress();
  const { container } = render(<RegisterAddress />);
  const submitValues = await fillValuesAndSubmit();
  expect(mockFn).toHaveBeenCalledWith(expect.objectContaining(submitValues));
  expect(container).toMatchSnapshot();
});
*/