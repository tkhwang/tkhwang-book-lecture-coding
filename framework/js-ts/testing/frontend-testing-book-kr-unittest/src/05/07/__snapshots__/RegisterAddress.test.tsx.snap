// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot: 등록 폼이 표시된다 1`] = `
<div>
  <div>
    <form>
      <h2>
        배송지 정보 입력
      </h2>
      <fieldset>
        <legend>
          연락처
        </legend>
        <div>
          <label>
            전화번호
            <input
              name="phoneNumber"
              type="text"
            />
          </label>
        </div>
        <div>
          <label>
            이름
            <input
              name="name"
              type="text"
            />
          </label>
        </div>
      </fieldset>
      <fieldset>
        <legend>
          배송지
        </legend>
        <div>
          <label>
            우편번호
            <input
              name="postalCode"
              placeholder="16397"
              type="text"
            />
          </label>
        </div>
        <div>
          <label>
            시/도
            <input
              name="prefectures"
              placeholder="경기도"
              type="text"
            />
          </label>
        </div>
        <div>
          <label>
            시/군/구
            <input
              name="municipalities"
              placeholder="수원시 권선구"
              type="text"
            />
          </label>
        </div>
        <div>
          <label>
            도로명
            <input
              name="streetNumber"
              placeholder="매곡로 67"
              type="text"
            />
          </label>
        </div>
      </fieldset>
      <hr />
      <div>
        <button>
          주문내용 확인
        </button>
      </div>
    </form>
  </div>
</div>
`;
